common/src/View/ClipTool.cpp

```diff
@@ -832,14 +832,19 @@ namespace TrenchBroom {
         }
 
         void ClipTool::addBrushesToRenderer(const std::map<Model::Node*, std::vector<Model::Node*>>& map, Renderer::BrushRenderer& renderer) {
-            Model::CollectBrushesVisitor collect;
+            std::vector<Model::BrushNode*> brushes;
 
             for (const auto& entry : map) {
-                const auto& brushes = entry.second;
-                Model::Node::accept(std::begin(brushes), std::end(brushes), collect);
+                const auto& nodes = entry.second;
+                for (auto* node : nodes) {
+                    node->acceptLambda(kdl::overload {
+                        [&](Model::BrushNode* brush) { brushes.push_back(brush); },
+                        [](auto*) {}
```

Comment:
General question about these cases, how does this `auto*` compile? What is the type inferred as? I guess the only answer  that makes sense is `Model::Node*`, but it's kind of surprising that the compiler can figure that out.

Reply:
I don't know what the compiler makes of this. It could generate overloads for all types it encounters (which I would assume it does since the types might not be related at all), or it could figure out the common type (which it already can, see `std::common_type`).

Reply:
>  It could generate overloads for all types it encounters (which I would assume it does since the types might not be related at all),

The `auto*` doesn't make it a template afaik, so that doesn't sound possible to me. I'm expecting it to compile as 1 lambda. But, we can check what it's doing I guess.

We should probably have a test case that this pattern actually does what we expect.

Reply:
Haha, of course it does neither of those things. `[](auto*) {}` generates a struct with a templated call operator:

```
struct blah {
    template <typename T>
    void operator()(T*) {}
};
```

The struct generated by `kdl::overload` then inherits that template, but it only matches if the argument type is not `Model::BrushNode*`. In particular, this means that if you do this:

```
node->acceptLambda(kdl::overload {
    [](Model::Object* object { ... },
    [](auto*) {}
```

it won't match anything because `Model::Object` is abstract, and a call with `Model::BrushNode*` matches the templated call operator better. So it can be tricky when using `auto*`. We can also ban this by convention if you prefer. I try to use it only in situations where it seems reasonable to assume that 1) it does the right thing and 2) it doesn't silently break in the future if we add new node types.

Reply:
Ok, thanks, that makes more sense. I did some testing and this `struct blah` behaves the same as the lambda version below it.

```
        struct blah {
            template <typename T>
            void operator()(T* wildcard) const {
                    std::cout << "wildcard: " << wildcard->name() << " type: " << typeid(decltype(wildcard)).name() << "\n";
            }
            void operator()(Model::BrushNode* brush) const {
                std::cout << "brush: " << brush->name() << "\n";
            }
        };

...

        // nodes contains a group, a brush, a world

            std::cout << "lambda version:\n";
            for (Model::Node* node : nodes) {
                node->acceptLambda(kdl::overload{
                    [](auto* wildcard) {
                        std::cout << "wildcard: " << wildcard->name() << " type: " << typeid(decltype(wildcard)).name() << "\n";
                    },
                    [](Model::BrushNode* brush) {
                        std::cout << "brush: " << brush->name() << "\n";
                    },
                });
            }

            std::cout << "struct version:\n";
            for (Model::Node* node : nodes) {
                node->acceptLambda(blah{});
            }
```

Reply:
> We can also ban this by convention if you prefer. I try to use it only in situations where it seems reasonable to assume that 1) it does the right thing and 2) it doesn't silently break in the future if we add new node types.

I'm OK with the `auto*` now that I know what's going on :-)

It makes sense to use it in cases where you want to e.g. "do something if node is a BrushNode" (effectively a dynamic_cast) and you don't care about the other types.